Explain domain service and its usage in command handler (#963)
PrestaShop/docs: src/content/1.7/development/architecture/domain/services.md

---
title: Domain services
weight: 20
---

# Domain services

> As mentioned in [CQRS principles](cqrs.md), a single command is just a port, a representation of a single use case, therefore the command handler should not contain any domain logic by itself, but should be orchestrating other services to fulfill the use case instead.

Reviewer: ```suggestion
As mentioned in [CQRS principles]({{< relref "cqrs" >}}), a Command is a representation of a single domain use case. To prevent Command handlers from duplicating code and from depending on each other, they usually don't implement the actual logic themselves, but delegate it to Domain Services.
```

These services are responsible for logic such as:
1. Persisting the entities (a.k.a ObjectModel in current prestashop architecture) data into database
2. Validating the entities
3. Performing various modifications of related data structures etc.

> Here are couple principles of the domain services: 

Reviewer: ```suggestion
Here are some principles of the domain services: 
```

> 1. It MUST use the existing `ObjectModel` for writes as long as those classes exist.

Reviewer: ```suggestion
1. They MUST use the existing `ObjectModel` for writes, as long as those classes exist.
```

> 2. If it uses any legacy service or object model it MUST stay in adapter namespace.

Reviewer: ```suggestion
2. If they uses any legacy service or object model then they MUST be placed in the `Adapter` namespace.
```

> 3. If it uses any `ObjectModel` method related to data persistence, that method MUST be wrapped into a dedicated repository class which must ensure that no legacy exceptions are thrown (some reusable methods are present in [AbstractObjectModelRepository](https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/AbstractObjectModelRepository.php))

Reviewer: ```suggestion
3. If they need to perform a query and a related `ObjectModel` exists, then this query MUST be delegated to the appropriate repository class, which must ensure that no legacy exceptions are thrown (some reusable methods are present in [AbstractObjectModelRepository](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Adapter/AbstractObjectModelRepository.php)). If the related `ObjectModel` already implements this method, then the repository must delegate its implementation to the `ObjectModel`.
```

> 4. If `ObjectModel` is used for any write operation it MUST be validated by a dedicated validator class (some reusable methods are present in [AbstractObjectModelValidator](https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/AbstractObjectModelValidator.php))

Reviewer: Don't use `develop` for links, choose a branch or a tag, because the content might be deleted in a future commit.
```suggestion
4. If `ObjectModel` is used for any write operation it MUST be validated by a dedicated validator class (some reusable methods are present in [AbstractObjectModelValidator](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Adapter/AbstractObjectModelValidator.php))
```

---
title: Domain services

weight: 20
---

# Domain services

> As mentioned in [CQRS principles](../cqrs), a `Command` is a representation of a single domain use case. To
> prevent `Command Handlers` from duplicating the code and depending on each other, it is usually best to **avoid keeping
> the actual logic in the `Command Handler`** itself, but delegate it to the `Domain Services`. These services are responsible for
> logic such as:

Reviewer: Avoid breaking paragraphs, use soft wraps in your editor as necessary
```suggestion
As mentioned in [CQRS principles]({{< relref "/cqrs" >}}), a `Command` is a representation of a single domain use case. To prevent `Command Handlers` from duplicating code and depending on each other, it is usually best to **avoid implementing the actual logic in the `Command Handler`** itself, and delegate it to `Domain Services` instead. These services are responsible for logic such as:
```

1. Persisting the entities (a.k.a ObjectModel in current prestashop architecture) data into database
2. Validating the entities
3. Performing various modifications of related data structures etc.

> Here are some principles of the domain service implementation:

Reviewer: ```suggestion
Here are some principles for implementing a Domain Service:
```

> 1. It MUST use the existing `ObjectModel` for writes, as long as those classes exist.

Reviewer: ```suggestion
1. It MUST use an existing `ObjectModel` for writes, as long as such class exists.
```

2. If it uses any legacy service or object model, then it MUST be placed in the `Adapter` namespace.

> 3. If it needs to perform a sql query and related `ObjectModel` exists, then this query MUST be delegated to the
>    appropriate repository class which must ensure that no legacy exceptions are thrown (some reusable methods are
>    present
>    in [AbstractObjectModelRepository](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Adapter/AbstractObjectModelRepository.php))
>    . If the related `ObjectModel` already implements such method, then the repository must delegate its implementation
>    to the `ObjectModel`.

Reviewer: Don't break paragraphs
```suggestion
3. If it needs to perform an SQL query and a related `ObjectModel` exists, then this query MUST be delegated to the appropriate repository class which must ensure that no legacy exceptions are thrown (some reusable methods are present in [AbstractObjectModelRepository(https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Adapter/AbstractObjectModelRepository.php)). If the related `ObjectModel` already implements such method, then the repository must delegate its implementation to the `ObjectModel`.
```

Author: yeah I suppose I cannot use phpstorm for cleaning this then :smile:

> 4. If `ObjectModel` is used for any write operation it MUST be validated by a dedicated validator class (some reusable
>    methods are present
>    in [AbstractObjectModelValidator](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Adapter/AbstractObjectModelValidator.php)).

Reviewer: ```suggestion
4. If `ObjectModel` is used for any write operation, then it MUST be validated by a dedicated validator class (some reusable methods are present in [AbstractObjectModelValidator](https://github.com/PrestaShop/PrestaShop/blob/1.7.8.x/src/Adapter/AbstractObjectModelValidator.php)).
```

Reviewer: Should or must?

Author: hmm "MUST ... if there are any validation rules in related `ObjectModel`" ?
Its a must for repository to wrap the validation errors, else we will let the PrestashopException to bubble up.
Makes sense?
